src/ert/ensemble_evaluator/dispatch.py

```diff
@@ -74,34 +60,59 @@ def done_logger(_):
         events_handling.add_done_callback(done_logger)
         await events_handling
 
+    def run_dispatcher(self, loop):
+        try:
+            loop.run_until_complete(self._job())
+        except asyncio.CancelledError as ex:
+            logger.warning(f"batcher was cancelled: {ex}")
+        except Exception as failure:  # pylint: disable=broad-exception-caught
+            logger.exception(f"exception in batching dispatcher: {failure}")
+            trace_info = traceback.format_exception(
+                type(failure), failure, failure.__traceback__
+            )
+            logger.error(f"{trace_info}")
+        else:
+            logger.debug("batcher finished normally")
+            return
+        finally:
+            self._finished = True
+        loop.run_until_complete(self._done_callback())
+        logger.debug("Dispatcher thread exiting.")
+
+    async def _done_callback(self):
+        # call any registered handlers for FAILED. since we don't have
+        # an event, pass empty list and let handler decide how to proceed
+        funcs = self._LOOKUP_MAP[identifiers.EVTYPE_ENSEMBLE_FAILED]
+        await asyncio.gather(*[f([]) for f, _ in funcs])
+
     async def _job(self):
         while self._running:
-            await asyncio.sleep(self._timeout)
+            if len(self._buffer) < self._max_batch:
+                time.sleep(self._sleep_between_batches_seconds)
+            else:
+                time.sleep(0)
             await self._work()
-
         # Make sure no events are lingering
         await self._work()
 
     async def join(self):
         self._running = False
         # if result is exception it should have been handled by
         # done-handler, but also avoid killing the caller here
-        with contextlib.suppress(BaseException):
-            await self._task
+        while not self._finished:
+            await asyncio.sleep(0.01)
 
-    def register_event_handler(self, event_types, function, batching=True):
+    def register_event_handler(self, event_types, function):
         if not isinstance(event_types, set):
             event_types = {event_types}
         for event_type in event_types:
-            self._LOOKUP_MAP[event_type].append((function, batching))
+            self._LOOKUP_MAP[event_type].append(function)
 
     async def handle_event(self, event):
-        for function, batching in self._LOOKUP_MAP[event["type"]]:
-            if batching:
-                if self._task.done():
-                    raise asyncio.InvalidStateError(
-                        "trying to handle event after batcher is done"
-                    )
+        for function in self._LOOKUP_MAP[event["type"]]:
+            if not self._running:
```

Comment:
I guess this can go above the `for loop`

Reply:
fixed in 811e1af82fdc61eb75f7772aca114793dc997448

Reply:
i also took the liberty, after looking at that code, of [simplifying further](https://github.com/equinor/ert/pull/5828/commits/81b0aa0ecfca33959d49b5257fb026945cd99405) - check it out and let me know what you think